File: src/px4_ros_com/src/detection.cpp
#include <memory>
#include <vector>
#include <string>
#include <sstream>

// ROS2
#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/image.hpp"
#include "cv_bridge/cv_bridge.h"

// OpenCV
#include "opencv2/opencv.hpp"

// ------------------------------
// Detector Interface
// ------------------------------
class Detector
{
public:
  struct DetectionResult {
    cv::Rect bounding_box;
    std::string label;
  };

  virtual ~Detector() = default;
  virtual std::vector<DetectionResult> detect(const cv::Mat & image) = 0;
};

// ------------------------------
// RedDetector Implementation (Detects Any Red Object)
// ------------------------------
class RedDetector : public Detector
{
public:
  std::vector<DetectionResult> detect(const cv::Mat & image) override
  {
    std::vector<DetectionResult> detections;

    RCLCPP_INFO(rclcpp::get_logger("RedDetector"), "Starting red detection process...");

    // 1. Convert from BGR to HSV color space for easier thresholding.
    cv::Mat hsv;
    cv::cvtColor(image, hsv, cv::COLOR_BGR2HSV);
    RCLCPP_DEBUG(rclcpp::get_logger("RedDetector"), "Converted image to HSV.");
  

    // 2. Threshold HSV image to get only red colors.
    cv::Mat mask1, mask2;
    cv::inRange(hsv, cv::Scalar(0, 120, 70),  cv::Scalar(10, 255, 255), mask1);
    cv::inRange(hsv, cv::Scalar(170, 120, 70), cv::Scalar(180, 255, 255), mask2);
    cv::Mat mask = mask1 | mask2;
    RCLCPP_DEBUG(rclcpp::get_logger("RedDetector"), "Created raw red mask.");
  

    // 3. Clean the mask using morphological operations.
    cv::Mat kernel = cv::getStructuringElement(cv::MORPH_RECT, cv::Size(3, 3));
    cv::morphologyEx(mask, mask, cv::MORPH_OPEN, kernel);
    cv::morphologyEx(mask, mask, cv::MORPH_DILATE, kernel);
    RCLCPP_DEBUG(rclcpp::get_logger("RedDetector"), "Cleaned the red mask.");

    // 4. Detect contours in the cleaned mask.
    std::vector<std::vector<cv::Point>> contours;
    cv::findContours(mask, contours, cv::RETR_EXTERNAL, cv::CHAIN_APPROX_SIMPLE);
    RCLCPP_DEBUG(rclcpp::get_logger("RedDetector"), "Found %zu contours.", contours.size());

    // 5. Process each contour: if it is large enough, consider it a red object.
    for (const auto & contour : contours)
    {
      double area = cv::contourArea(contour);
      if (area > 500)  // Tune this threshold as necessary.
      {
        cv::Rect bbox = cv::boundingRect(contour);
        DetectionResult result;
        result.bounding_box = bbox;
        result.label = "Red Object";
        detections.push_back(result);

        // Log the detection details.
        std::stringstream ss;
        ss << "Red object detected at (x=" << bbox.x << ", y=" << bbox.y 
           << ", w=" << bbox.width << ", h=" << bbox.height << ")";
        RCLCPP_INFO(rclcpp::get_logger("RedDetector"), "%s", ss.str().c_str());
      }
    }

    return detections;
  }
};

// ------------------------------
// Image Listener Node (ROS2 Node)
// ------------------------------
class ImageListenerNode : public rclcpp::Node
{
public:
  ImageListenerNode(std::shared_ptr<Detector> detector)
  : Node("image_listener_node"), detector_(detector)
  {
    // Subscribe to the image topic (update the topic name if needed).
    subscription_ = this->create_subscription<sensor_msgs::msg::Image>(
      "/world/default/model/x500_mono_cam_down_0/link/camera_link/sensor/imager/image", 10,
      std::bind(&ImageListenerNode::image_callback, this, std::placeholders::_1));
  }

private:
  void image_callback(const sensor_msgs::msg::Image::SharedPtr msg)
  {
    RCLCPP_INFO(this->get_logger(), "Received new image frame.");

    // Convert the ROS image message to an OpenCV image.
    cv_bridge::CvImagePtr cv_ptr;
    try {
      cv_ptr = cv_bridge::toCvCopy(msg, "bgr8");
    } catch (cv_bridge::Exception &e) {
      RCLCPP_ERROR(this->get_logger(), "cv_bridge exception: %s", e.what());
      return;
    }
    cv::Mat frame = cv_ptr->image;

    // Run detection.
    std::vector<Detector::DetectionResult> results = detector_->detect(frame);

    // Draw bounding boxes and labels on the image.
    for (const auto & detection : results)
    {
      cv::rectangle(frame, detection.bounding_box, cv::Scalar(0, 255, 0), 2);
      cv::putText(frame, detection.label,
                  cv::Point(detection.bounding_box.x, detection.bounding_box.y - 10),
                  cv::FONT_HERSHEY_SIMPLEX, 0.5, cv::Scalar(0, 255, 0), 2);
    }

    // Show the final image with detections.
    cv::imshow("Final Detections", frame);
    cv::waitKey(1);
  }

  rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr subscription_;
  std::shared_ptr<Detector> detector_;
};

// ------------------------------
// Main Function
// ------------------------------
int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);

  // Create an instance of the red detector.
  std::shared_ptr<Detector> detector = std::make_shared<RedDetector>();

  // Create and spin the image listener node with the detector.
  auto node = std::make_shared<ImageListenerNode>(detector);
  rclcpp::spin(node);

  cv::destroyAllWindows();  // Cleanup OpenCV windows.
  rclcpp::shutdown();
  return 0;
}

File: src/px4_ros_com/src/detection.py
#!/usr/bin/env python3
"""
ROS2 Detection Node

Key Behaviors:
  1. Modularity:
       - Swap out `MultiColorDetector` for any other `Detector` implementation
         (e.g. a deep-learning model) without changing the node logic.
  2. Timed Execution:
       - Detection runs only once every 5 seconds, regardless of camera frame rate.
  3. Data Structures:
       - `detection_labels_` holds the list of labels detected.
       - `detection_coordinates_` holds the corresponding bounding boxes.
  4. Logging & Visualization:
       - All detections are logged via ROS2 logger.
       - Results are drawn on the frame and shown in an OpenCV window.
"""

import rclpy
from rclpy.node import Node

from sensor_msgs.msg import Image
from cv_bridge import CvBridge, CvBridgeError

import cv2
import numpy as np
from abc import ABC, abstractmethod
from typing import List, Tuple


class DetectionResult:
    """Holds one detection’s bounding box and label."""
    def __init__(self, bbox: Tuple[int, int, int, int], label: str):
        self.bounding_box = bbox  # (x, y, w, h)
        self.label = label


class Detector(ABC):
    """Abstract detector interface."""
    @abstractmethod
    def detect(self, image: np.ndarray) -> List[DetectionResult]:
        ...

class MultiColorDetector(Detector):
    """Detects red, green, blue objects via HSV thresholding."""
    def __init__(self):
        # (name, lower HSV, upper HSV)
        self.thresholds = [
            ("Red Object",   np.array([0, 120,  70]),  np.array([10, 255, 255])),
            ("Green Object", np.array([36,  50,  70]),  np.array([89, 255, 255])),
            ("Blue Object",  np.array([90,  50,  70]),  np.array([128, 255, 255])),
        ]
        # 3×3 rectangular kernel
        self.kernel = cv2.getStructuringElement(cv2.MORPH_RECT, (3, 3))

    def detect(self, image: np.ndarray) -> List[DetectionResult]:
        detections: List[DetectionResult] = []
        hsv = cv2.cvtColor(image, cv2.COLOR_BGR2HSV)

        for name, lower, upper in self.thresholds:
            mask = cv2.inRange(hsv, lower, upper)
            # clean up
            mask = cv2.morphologyEx(mask, cv2.MORPH_OPEN,  self.kernel)
            mask = cv2.morphologyEx(mask, cv2.MORPH_DILATE, self.kernel)

            contours, _ = cv2.findContours(mask, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
            for cnt in contours:
                if cv2.contourArea(cnt) > 500:
                    x, y, w, h = cv2.boundingRect(cnt)
                    detections.append(DetectionResult((x, y, w, h), name))

        return detections


class ImageListenerNode(Node):
    """ROS2 node that subscribes to Image, runs detection every 5 seconds."""
    def __init__(self, detector: Detector):
        super().__init__('image_listener_node')
        self.bridge = CvBridge()
        self.detector = detector
        self.latest_frame = None

        self.detection_labels_: List[str] = []
        self.detection_coordinates_: List[Tuple[int,int,int,int]] = []

        # subscribe to camera topic
        self.create_subscription(
            Image,
            '/camera',
            self.image_callback,
            10
        )

        # timer fires every 5 seconds, regardless of frame rate
        self.create_timer(2.0, self.timer_callback)

    def image_callback(self, msg: Image):
        """Save the latest frame for timed processing."""
        try:
            cv_img = self.bridge.imgmsg_to_cv2(msg, 'bgr8')
        except CvBridgeError as e:
            self.get_logger().error(f'CvBridge error: {e}')
            return

        self.latest_frame = cv_img.copy()
        cv2.waitKey(1)

    def timer_callback(self):
        """Called once every 5 seconds to run detection + logging + display."""
        if self.latest_frame is None:
            self.get_logger().warn('No frame available for detection.')
            return

        # run detector
        results = self.detector.detect(self.latest_frame)


        # annotate results
        for det in results:
            x, y, w, h = det.bounding_box
            self.detection_labels_.append(det.label)
            self.detection_coordinates_.append(det.bounding_box)

            cv2.rectangle(self.latest_frame, (x, y), (x+w, y+h), (0, 255, 0), 2)
            cv2.putText(
                self.latest_frame, det.label,
                (x, y-10), cv2.FONT_HERSHEY_SIMPLEX,
                0.5, (0, 255, 0), 2
            )
            self.get_logger().info(
                f"After detection {det.label}: "
                f"labels={self.detection_labels_}, "
                f"boxes={self.detection_coordinates_}"
            )

        # log labels
        label_str = ', '.join(self.detection_labels_) or 'None'
        self.get_logger().info(f'Detected labels: {label_str}')

        # log coordinates
        coord_str = ', '.join(f'({x},{y},{w},{h})' for x,y,w,h in self.detection_coordinates_) or 'None'
        self.get_logger().info(f'Coordinates: {coord_str}')

        # display
        cv2.imshow('Detections', self.latest_frame)
        cv2.waitKey(1)


def main(args=None):
    rclpy.init(args=args)
    detector = MultiColorDetector()
    node = ImageListenerNode(detector)
    try:
        rclpy.spin(node)
    finally:
        node.destroy_node()
        cv2.destroyAllWindows()
        rclpy.shutdown()


if __name__ == '__main__':
    main()
